Regression in calling `UIAElementInfo.descendants` with `depth` and criteria like `control_type` arguments

While working on #1334, I had questions about the implementation of `UIAElementInfo.descendants`, so I conducted an investigation.

## Expected Behavior
The following tests I added in `pywinauto/unittests/test_uia_element_info.py` should be passing. These tests passed when I added them while checking out [`0.6.8`](https://github.com/pywinauto/pywinauto/tree/0.6.8).

```py
# ...
# if UIA_support:
#     class UIAElementInfoTests(unittest.TestCase):
        # ...
        def test_default_depth_with_criteria_descendants(self):
            # There is no element with `control_type='Slider'` in children of the dialog.
            self.assertGreater(len(self.dlg.descendants(control_type="Slider")), len(self.dlg.children(control_type="Slider")))

        def test_depth_level_one_with_criteria_descendants(self):
            self.assertEqual(self.dlg.children(control_type="ToolBar"), self.dlg.descendants(control_type="ToolBar", depth=1))
            self.assertEqual(self.dlg.children(control_type="Button"), self.dlg.descendants(control_type="Button", depth=1))

        def test_depth_level_three_with_criteria_descendants(self):
            slider = next(iter(self.ctrl.descendants(control_type="Slider", depth=3)), None)
            self.assertIsNotNone(slider)
            self.assertIn(slider, self.ctrl.descendants(depth=3))
        # ...
```

## Actual Behavior
Adding the above-mentioned tests to the current [`atspi`](https://github.com/pywinauto/pywinauto/tree/atspi) branch(bf7f789d01) does **not** result in them passing.

```
======================================================================
FAIL: test_depth_level_three_with_criteria_descendants (pywinauto.unittests.test_uia_element_info.UIAElementInfoTests.test_depth_level_three_with_criteria_descendants)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "...\pywinauto\pywinauto\unittests\test_uia_element_info.py", line 131, in test_depth_level_three_with_criteria_descendants
    self.assertIsNotNone(slider)
AssertionError: unexpectedly None
```

## Steps to Reproduce the Problem

Please see above.

## Short Example of Code to Demonstrate the Problem

Please see above.

## Specifications

  - Pywinauto version:  the current [`atspi`](https://github.com/pywinauto/pywinauto/tree/atspi) branch(bf7f789d01)
  - Python version and bitness: Python 3.11.4 64bit
  - Platform and OS: Windows10 64-bit

### Appendix
`iter_descendants` still has an issue where it doesn't produce the same results as `descendants` when it took criteria args, even as of version `0.6.8`. I plan to address this after #1334 is merged.

--- a/pywinauto/element_info.py
+++ b/pywinauto/element_info.py
@@ -166,21 +166,6 @@ class ElementInfo(object):
         else:
             return elements
 
-    def get_descendants_with_depth(self, depth=None, **kwargs):
-        """Return a list of all descendant children of the element with the specified depth"""
-        descendants = []
-
-        def walk_the_tree(root, depth, **kwargs):
-            if depth == 0:
-                return
-            for child in root.children(**kwargs):
-                descendants.append(child)
-                next_depth = None if depth is None else depth - 1
-                walk_the_tree(child, next_depth, **kwargs)
-
-        walk_the_tree(self, depth, **kwargs)
-        return descendants
-
     def descendants(self, **kwargs):
         """Return descendants of the element"""
         raise NotImplementedError()

--- a/pywinauto/unittests/test_uia_element_info.py
+++ b/pywinauto/unittests/test_uia_element_info.py
@@ -118,6 +118,35 @@ if UIA_support:
             """Test whether a list of descendants with invalid depth raises exception"""
             self.assertRaises(Exception, self.ctrl.descendants, depth='qwerty')
 
+        def test_default_depth_with_criteria_descendants(self):
+            """Test whether a list of descendants with default depth and same criteria of the element is equal"""
+            self.assertEqual(
+                len(self.ctrl.descendants(depth=None, control_type="RadioButton")),
+                len(self.ctrl.descendants(control_type="RadioButton")),
+            )
+
+        def test_default_and_specified_depth_level_with_criteria_descendants(self):
+            """Test whether specifying a depth level will find fewer elements than without specifying it"""
+            self.assertLess(
+                len(self.ctrl.descendants(control_type="Text", depth=2)),
+                len(self.ctrl.descendants(control_type="Text")),
+            )
+
+        def test_depth_level_one_with_criteria_descendants(self):
+            """Test whether a list of descendants with depth=1 and same criteria of the element is equal to children"""
+            self.assertEqual(
+                self.ctrl.children(control_type="ToolBar"), self.ctrl.descendants(control_type="ToolBar", depth=1)
+            )
+
+        def test_depth_level_more_than_one_with_criteria_descendants(self):
+            """Test whether an element not found with smaller depth levels is found with larger levels"""
+            # There is no element with `control_type='ScrollBar'` with depth level less than three
+            nothing = next(iter(self.ctrl.descendants(control_type="Slider", depth=2)), None)
+            self.assertIsNone(nothing)
+            slider = next(iter(self.ctrl.descendants(control_type="Slider", depth=3)), None)
+            self.assertIsNotNone(slider)
+            self.assertIn(slider, self.ctrl.descendants())
+
         def test_descendants_generator(self):
             """Test whether descendant generator iterates over correct elements"""
             descendants = [desc for desc in self.ctrl.iter_descendants(depth=3)]

--- a/pywinauto/windows/uia_element_info.py
+++ b/pywinauto/windows/uia_element_info.py
@@ -480,13 +480,31 @@ class UIAElementInfo(ElementInfo):
         if UIAElementInfo.use_raw_view_walker:
             return list(self.iter_descendants(**kwargs))
         else:
+            cache_enable = kwargs.pop('cache_enable', False)
             depth = kwargs.pop('depth', None)
             if depth is None:
-                cache_enable = kwargs.pop('cache_enable', False)
                 cond = IUIA().build_condition(**kwargs)
                 return self._get_elements(IUIA().tree_scope["descendants"], cond, cache_enable)
+            if depth == 1:
+                return self.children(cache_enable=cache_enable, **kwargs)
             else:
-                return self.get_descendants_with_depth(depth=depth, **kwargs)
+                if kwargs:
+                    cond = IUIA().build_condition(**kwargs)
+                    elements = self._get_elements(IUIA().tree_scope["descendants"], cond, cache_enable)
+                    return ElementInfo.filter_with_depth(elements, self, depth)
+                else:
+                    descendants = []
+
+                    def walk_the_tree(parent, depth_):
+                        # type: (UIAElementInfo, int) -> None
+                        if depth_ == 0:
+                            return
+                        for child in parent.children(cache_enable=cache_enable):
+                            descendants.append(child)
+                            walk_the_tree(child, depth_ - 1)
+
+                    walk_the_tree(self, depth)
+                    return descendants
 
     @property
     def visible(self):

--- a/pywinauto/windows/uia_element_info.py
+++ b/pywinauto/windows/uia_element_info.py
@@ -462,7 +462,7 @@ class UIAElementInfo(ElementInfo):
         """Return a list of only immediate children of the element
 
          * **kwargs** is a criteria to reduce a list by process,
-           class_name, control_type, content_only and/or title.
+           class_name, control_type, content_only and/or name.
         """
         if UIAElementInfo.use_raw_view_walker:
             return list(self.iter_children(**kwargs))
@@ -475,7 +475,7 @@ class UIAElementInfo(ElementInfo):
         """Return a list of all descendant children of the element
 
          * **kwargs** is a criteria to reduce a list by process,
-           class_name, control_type, content_only and/or title.
+           class_name, control_type, content_only and/or name.
         """
         if UIAElementInfo.use_raw_view_walker:
             return list(self.iter_descendants(**kwargs))

